src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -1195,86 +1237,111 @@ int fx_muxer_t::read_config_and_execute(
         trace::error(_X("The specified deps.json [%s] does not exist"), deps_file.c_str());
         return StatusCode::InvalidArgFailure;
     }
-    if (!runtime_config.empty() && (!pal::realpath(&runtime_config) || !pal::file_exists(runtime_config)))
-    {
-        trace::error(_X("The specified runtimeconfig.json [%s] does not exist"), runtime_config.c_str());
-        return StatusCode::InvalidConfigFile;
-    }
 
-    pal::string_t config_file, dev_config_file;
+    // Read config
+    // The fx_definitions_owned variable controls object lifetime
```

Comment:
Review note: unique_ptr is a smart pointer which will call delete when popped off the stack. It is not possible to pass std::vector<std::unique_ptr<fx_definition_t>> to other methods because unique_ptr can't be copied. Also we don't want to use shared_ptr instead (which does allow the pointer to be shared) as the ownership is simply stack-based. So another vector was created to pass around the raw pointers (fx_definitions). Another variant would be just to use fx_definitions (not fx_definitions_owned) and call delete[] when the method is done, but that is a bit more complex and error prone due to exception etc.

Reply:
May be best to talk in person to make sure we're on the same page, but here goes:
1. That means we would have to call delete() on the vector, correct? Also, when using an indexer on a vector of unique pointers, the copy ctor is attempted to be called on the unique pointer, which by design isn't supported which results in a compiler error
2. Move the vector from [where?] to corehost_init_t? Note that the array is also passed around elsewhere as reference information (no ownership)
3. We need to dynamically create the fx_definitions (we don't know how many there are), but yes we could create them, copy by value, and then call delete on the original immediately. There are also embedded types (runtime_config_t, deps_json_t) that would also need to make sure support default copy ctor (or add one). Doesn't seem as efficient.

Reply:
> That means we would have to call delete() on the vector, correct?

No, but it does means that the thing that owns the vector must be kept alive at least as long as corehost_init_t, which can be non-obvious so (1) probably isn't the best solution.

> Note that the array is also passed around elsewhere as reference information (no ownership)

I don't see why that would be a problem.

> Also, when using an indexer on a vector of unique pointers, the copy ctor is attempted to be called on the unique pointer, which by design isn't supported which results in a compiler error.

What gives you this impression? `vector::operator[]`  returns a reference.

Below is a demo of (2) with pass-by-reference and indexing too.

``` C++
#include <vector>
#include <memory>
#include <iostream>

using namespace std;

class fx_definition_t
{
public:
  fx_definition_t(const string& name) : m_name(name) 
  {
    cout << "construct fx_definition_t " << m_name.c_str() << endl; 
  }

  ~fx_definition_t() 
  { 
    cout << "destruct fx_definition_t" << m_name.c_str() << endl; 
  }

  const string& name() const { return m_name;  }

private:
  string m_name;
};

typedef vector<unique_ptr<fx_definition_t>> fx_definition_vector_t;

class corehost_init_t 
{
public:
  corehost_init_t(fx_definition_vector_t&& fx_definitions)
    : m_fx_definitions(std::move(fx_definitions)) 
  {
    cout << "construct corehost_init_t" << endl;
  }

  ~corehost_init_t()
  {
    cout << "destruct corehost_init_t" << endl;
  }

  const fx_definition_vector_t& fx_definitions() const 
  { 
    return m_fx_definitions; 
  }

private:
  fx_definition_vector_t m_fx_definitions;
};

// can pass vector of unique_ptr by reference
void dump_fx_definitions(const fx_definition_vector_t& fx_definitions)
{
  // can index into vector of unique_ptr
  for (size_t i = 0; i < fx_definitions.size(); i++)
  {
    cout << "visit fx_definition_t " << fx_definitions[i]->name().c_str() << endl;
  }
}

void execute_app(const corehost_init_t& init)
{
  dump_fx_definitions(init.fx_definitions());
}

int read_config_and_execute()
{
  fx_definition_vector_t fx_definitions;
  fx_definitions.emplace_back(new fx_definition_t("A"));
  fx_definitions.emplace_back(new fx_definition_t("B"));

  {
    corehost_init_t init(std::move(fx_definitions)); 
    execute_app(init);
  }

  cout << "init took over fx_definitions, so its empty here. size=" << fx_definitions.size() << endl;
  return 0;
}

int main() 
{
  return read_config_and_execute();
}
```

Reply:
Thanks for the detailed reply. The code now has a single vector, and there was no need to use std::move(). Passing the vector my reference was sufficient.

Reply:
Ah, for some reason I had in mind that corehost_init_t was going to hold on to the vector of pointers, which is where the move idea came from.